File: src/lib/tauri/serverAuth.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Snapshot of the desktop's binding to `waveflow-server` for the
 * Settings card. `url` is the base URL the desktop will hit for
 * `/api/v1/*` calls (app-wide). `signedIn` is `true` when the active
 * profile has a JWT row under the `waveflow_server` provider — it
 * does NOT validate the token against the server, which would cost an
 * HTTP round-trip the UI rarely needs.
 */
export interface ServerStatus {
  url: string | null;
  /**
   * waveflow-web URL the OAuth-loopback handshake (Phase
   * 1.f.desktop.1b) opens in the system browser. May differ from
   * `url` for deployments that proxy the API and the web on separate
   * domains.
   */
  web_url: string | null;
  signed_in: boolean;
}

/** Snapshot the server binding state. */
export function serverGetStatus(): Promise<ServerStatus> {
  return invoke<ServerStatus>("server_get_status");
}

/**
 * Persist the waveflow-server base URL. Empty string clears the row
 * (back to local-only mode). Validates parseability + http(s) scheme
 * server-side, so a typo surfaces as an error rather than silently
 * breaking future sync calls.
 */
export function serverSetUrl(url: string): Promise<ServerStatus> {
  return invoke<ServerStatus>("server_set_url", { url });
}

/** Persist the waveflow-web URL used by the OAuth-loopback flow. */
export function serverSetWebUrl(url: string): Promise<ServerStatus> {
  return invoke<ServerStatus>("server_set_web_url", { url });
}

/**
 * Persist the Bearer JWT the user pasted in from the browser
 * sign-in. Rejects empty / non-JWT-shaped input server-side.
 */
export function serverSetToken(token: string): Promise<ServerStatus> {
  return invoke<ServerStatus>("server_set_token", { token });
}

/** Drop the per-profile JWT. URL stays so the user can re-paste. */
export function serverSignOut(): Promise<ServerStatus> {
  return invoke<ServerStatus>("server_sign_out");
}

/**
 * Open the configured server URL in the user's default browser —
 * kept as a fallback for users who prefer to copy the JWT manually.
 * The OAuth-loopback flow below is the default since 1.f.desktop.1b.
 */
export function serverOpenLoginBrowser(): Promise<void> {
  return invoke<void>("server_open_login_browser");
}

/**
 * Run the local-loopback OAuth-style handshake against `waveflow-web`.
 * Opens the system browser to `<web-url>/desktop-login?cb=…&state=…`,
 * binds a one-shot listener on 127.0.0.1:49388 for up to three
 * minutes, and persists the JWT the web side hands back. Returns the
 * fresh `ServerStatus` once the round-trip completes.
 *
 * Throws on user cancellation, timeout, or `state` mismatch (treated
 * as a possible CSRF — the UI surfaces the error string verbatim).
 */
export function serverBeginLoopbackLogin(): Promise<ServerStatus> {
  return invoke<ServerStatus>("server_begin_loopback_login");
}

/**
 * Sync mode for the active profile. `local` skips the
 * sync_pending_op queue entirely even when a JWT is configured
 * (useful for privacy-conscious profiles); `hybrid` is the default
 * once signed in — reads stay local, writes hit local + the queue,
 * and the drain task posts them upstream.
 */
export type SyncMode = "local" | "hybrid";

export function syncGetMode(): Promise<SyncMode> {
  return invoke<SyncMode>("sync_get_mode");
}

export function syncSetMode(mode: SyncMode): Promise<SyncMode> {
  return invoke<SyncMode>("sync_set_mode", { req: { mode } });
}

File: src-tauri/crates/app/src/commands/server_auth.rs
//! Tauri commands for the `waveflow-server` account binding.
//!
//! Phase 1.f.desktop.1 — the foundational surface every later sub-PR
//! builds on. Sync (`1.f.desktop.2`), repo swap (`1.f.desktop.3`), and
//! WebSocket subscription (`1.f.desktop.4`) all read their config
//! through [`crate::server_client`].
//!
//! Sign-in flow today: the user opens the server URL in their
//! browser, signs into Better Auth, copies the JWT it issues, and
//! pastes it into the Settings card. A polished local-loopback OAuth
//! flow (mirroring `commands::spotify`) ships in `1.f.desktop.1b`
//! once the matching `/desktop-login` route lands on `waveflow-web`.

use crate::{
    error::{AppError, AppResult},
    server_client::{self, ServerStatus},
    state::AppState,
};

/// Snapshot of the desktop's server binding for the Settings card.
#[tauri::command]
pub async fn server_get_status(state: tauri::State<'_, AppState>) -> AppResult<ServerStatus> {
    server_client::status(&state).await
}

/// Persist the waveflow-server base URL. Validates the value parses
/// as `http(s)://…`; empty input clears the row so the desktop falls
/// back to local-only mode. App-wide (not per-profile) — see the
/// module docstring on [`crate::server_client`].
#[tauri::command]
pub async fn server_set_url(
    state: tauri::State<'_, AppState>,
    url: String,
) -> AppResult<ServerStatus> {
    server_client::write_url(&state.app_db, &url).await?;
    server_client::status(&state).await
}

/// Persist the Bearer JWT the user pasted in from the browser
/// sign-in. Rejects empty / structurally-bad input — leaving a broken
/// token in place would 401 every sync call. Per-profile.
#[tauri::command]
pub async fn server_set_token(
    state: tauri::State<'_, AppState>,
    token: String,
) -> AppResult<ServerStatus> {
    server_client::write_token(&state, &token).await?;
    server_client::status(&state).await
}

/// Sign out the active profile from waveflow-server. Idempotent — no
/// error when the user wasn't signed in to begin with. Server URL
/// stays in place (so the user can re-paste a fresh token without
/// re-entering the URL).
#[tauri::command]
pub async fn server_sign_out(state: tauri::State<'_, AppState>) -> AppResult<ServerStatus> {
    server_client::clear_token(&state).await?;
    server_client::status(&state).await
}

/// Open the persisted server URL in the user's default browser. The
/// landing page itself is whatever `waveflow-web` serves at the root
/// of the configured URL — typically the sign-in page when no
/// session is present, the dashboard otherwise. We don't append a
/// path because the post-1.f.desktop.1b flow will swap this for the
/// loopback handshake; until then the user navigates by hand.
///
/// Returns 400-style errors (via [`AppError::Other`]) when the URL
/// isn't configured yet, so the UI can pull the user back to the
/// "set URL" step instead of opening a blank tab.
#[tauri::command]
pub async fn server_open_login_browser(state: tauri::State<'_, AppState>) -> AppResult<()> {
    let url = server_client::read_url(&state.app_db)
        .await?
        .ok_or_else(|| {
            AppError::Other("server URL is not configured; set it in Settings first".into())
        })?;
    tauri_plugin_opener::open_url(url, None::<&str>)
        .map_err(|err| AppError::Other(format!("open server login: {err}")))?;
    Ok(())
}
